spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/JGitEnvironmentRepositoryTests.java

```diff
@@ -716,6 +727,111 @@ public void shouldSetTransportConfigCallbackOnCloneAndFetch() throws Exception {
+		when(git.status()).thenReturn(statusCommand);
+		Repository repository = mock(Repository.class);
+		when(git.getRepository()).thenReturn(repository);
+		StoredConfig storedConfig = mock(StoredConfig.class);
+		when(repository.getConfig()).thenReturn(storedConfig);
+		when(storedConfig.getString("remote", "origin", "url")).thenReturn("http://example/git");
+		when(statusCommand.call()).thenReturn(status);
+		when(status.isClean()).thenReturn(true);
+
+		// refresh()->fetch
+		FetchCommand fetchCommand = mock(FetchCommand.class);
+		FetchResult fetchResult = mock(FetchResult.class);
+
+		TrackingRefUpdate trackingRefUpdate = mock(TrackingRefUpdate.class);
+		Collection<TrackingRefUpdate> trackingRefUpdates = Collections.singletonList(trackingRefUpdate);
+
+		when(git.fetch()).thenReturn(fetchCommand);
+		when(fetchCommand.setRemote(anyString())).thenReturn(fetchCommand);
+		when(fetchCommand.call()).thenReturn(fetchResult);
+		when(fetchResult.getTrackingRefUpdates()).thenReturn(trackingRefUpdates);
+
+		// refresh()->deleteBranch
+		ReceiveCommand receiveCommand = mock(ReceiveCommand.class);
+		when(trackingRefUpdate.asReceiveCommand()).thenReturn(receiveCommand);
+		when(receiveCommand.getType()).thenReturn(ReceiveCommand.Type.DELETE);
+		when(trackingRefUpdate.getLocalName()).thenReturn("refs/remotes/origin/feature/deletedBranchFromOrigin");
+
+		DeleteBranchCommand deleteBranchCommand = mock(DeleteBranchCommand.class);
+		when(git.branchDelete()).thenReturn(deleteBranchCommand);
+		when(deleteBranchCommand.setBranchNames(eq("feature/deletedBranchFromOrigin"))).thenReturn(deleteBranchCommand);
+		when(deleteBranchCommand.setForce(true)).thenReturn(deleteBranchCommand);
+		when(deleteBranchCommand.call()).thenThrow(new NotMergedException());// here
+																			// is
+																			// our
+																			// exception
+																			// we
+																			// are
+																			// testing
+
+		// refresh()->checkout
+		CheckoutCommand checkoutCommand = mock(CheckoutCommand.class);
+		// refresh()->checkout->containsBranch
+		ListBranchCommand listBranchCommand = mock(ListBranchCommand.class);
+		when(git.checkout()).thenReturn(checkoutCommand);
+		when(git.branchList()).thenReturn(listBranchCommand);
+		List<Ref> refs = new ArrayList<>();
+		Ref ref = mock(Ref.class);
+		refs.add(ref);
+		when(ref.getName()).thenReturn("/master");
+		when(listBranchCommand.call()).thenReturn(refs);
+
+		// refresh()->merge
+		MergeResult mergeResult = mock(MergeResult.class);
+		MergeResult.MergeStatus mergeStatus = mock(MergeResult.MergeStatus.class);
+		MergeCommand mergeCommand = mock(MergeCommand.class);
+		when(git.merge()).thenReturn(mergeCommand);
+		when(mergeCommand.call()).thenReturn(mergeResult);
+		when(mergeResult.getMergeStatus()).thenReturn(mergeStatus);
+		when(mergeStatus.isSuccessful()).thenReturn(true);
+
+		// refresh()->return
+		// git.getRepository().findRef("HEAD").getObjectId().getName();
```

Comment:
You commented this out because it wont compile right?
I think it should be changed to `when(repository.getRef(anyString())).thenReturn(headRef);`
I am surprised this test even passed without it....so that is quite concerning

Reply:
actually this expectation is redundant, just take a look a bit lower and you will see actual mock
```when(repository.getRef(anyString())).thenReturn(headRef);``` 
As for test and how it passes - it's a big trouble of this class, since it's huge and getting bigger, so mock&verify correctly is very hard.